src/py/flwr/server/state/sqlite_state.py

```diff
@@ -0,0 +1,582 @@
+            """
+
+            # Prepare data for query
+            delivered_at = _now().isoformat()
+            data = {"delivered_at": delivered_at}
+            for index, task_id in enumerate(found_task_ids):
+                data[f"id_{index}"] = str(task_id)
+
+            # Run query
+            rows = self.query(query, data)
+
+        result = [dict_to_task_res(row) for row in rows]
+        return result
+
+    def num_task_ins(self) -> int:
+        """Number of task_ins in store.
+
+        This includes delivered but not yet deleted task_ins.
+        """
+        query = "SELECT count(*) AS num FROM task_ins;"
+        rows = self.query(query)
+        result = rows[0]
+        num = cast(int, result["num"])
+        return num
+
+    def num_task_res(self) -> int:
+        """Number of task_res in store.
+
+        This includes delivered but not yet deleted task_res.
+        """
+        query = "SELECT count(*) AS num FROM task_res;"
+        rows = self.query(query)
+        result: Dict[str, int] = rows[0]
+        return result["num"]
+
+    def delete_tasks(self, task_ids: Set[UUID]) -> None:
+        """Delete all delivered TaskIns/TaskRes pairs."""
+        ids = list(task_ids)
+        if len(ids) == 0:
+            return None
+
+        placeholders = ",".join([f":id_{index}" for index in range(len(task_ids))])
+        data = {f"id_{index}": str(task_id) for index, task_id in enumerate(task_ids)}
+
+        # 1. Query: Delete task_ins which have a delivered task_res
+        query_1 = f"""
+            DELETE FROM task_ins
+            WHERE delivered_at != ''
+            AND task_id IN (
+                SELECT ancestry
+                FROM task_res
+                WHERE ancestry IN ({placeholders})
+                AND delivered_at != ''
+            );
+        """
+
+        # 2. Query: Delete delivered task_res to be run after 1. Query
+        query_2 = f"""
+            DELETE FROM task_res
+            WHERE ancestry IN ({placeholders})
+            AND delivered_at != '';
+        """
+
+        if self.conn is None:
+            raise Exception("State not intitialized")
+
+        with self.conn:
+            self.conn.execute(query_1, data)
+            self.conn.execute(query_2, data)
+
+        return None
+
+    def register_node(self, node_id: int) -> None:
+        """Store `node_id` in state."""
+        query = "INSERT INTO node VALUES(:id);"
+        self.query(query, (node_id,))
+
+    def unregister_node(self, node_id: int) -> None:
+        """Remove `node_id` from state."""
+        query = "DELETE FROM node WHERE id = :id;"
+        self.query(query, (node_id,))
+
+    def get_nodes(self) -> Set[int]:
+        """Retrieve all currently stored node IDs as a set."""
+        query = "SELECT * FROM node;"
+        rows = self.query(query)
+        result: Set[int] = {row["id"] for row in rows}
+        return result
+
+
+def _now() -> datetime:
+    return datetime.now(tz=timezone.utc)
```

Comment:
I believe we have this somewhere else, too. Should we put it somewhere in `flwr.common`?

Reply:
doing it